pkg/syncer/resourcesync/controller.go

```diff
@@ -0,0 +1,330 @@
+	InformerForResource(gvr schema.GroupVersionResource) (*SyncerInformer, bool)
+	Start(ctx context.Context, numThreads int)
+}
+
+type ResourceEventHandlerPerGVR func(schema.GroupVersionResource) cache.ResourceEventHandler
+
+// controller is a control loop that watches synctarget. It starts/stops spec syncer and status syncer
+// per gvr based on synctarget.Status.SyncedResources.
+// All the spec/status syncer share the same downstreamNSInformer and upstreamSecretInformer. Informers
+// for gvr is started separated for each syncer.
+type Controller struct {
+	queue                        workqueue.RateLimitingInterface
+	upstreamDynamicClusterClient *dynamic.Cluster
+	downstreamDynamicClient      dynamic.Interface
+
+	upstreamEventHandlers   []ResourceEventHandlerPerGVR
+	downstreamEventHandlers []ResourceEventHandlerPerGVR
+
+	syncTargetName      string
+	syncTargetWorkspace logicalcluster.Name
+	syncTargetUID       types.UID
+	syncTargetLister    workloadlisters.SyncTargetLister
+
+	syncerInformerMap map[schema.GroupVersionResource]*SyncerInformer
+	mutex             sync.RWMutex
+}
+
+func NewController(
+	upstreamDynamicClusterClient *dynamic.Cluster,
+	downstreamDynamicClient dynamic.Interface,
+	syncTargetInformer workloadinformers.SyncTargetInformer,
+	syncTargetName string,
+	syncTargetWorkspace logicalcluster.Name,
+	syncTargetUID types.UID,
+) (SyncerInformerFactory, error) {
+	c := &Controller{
+		queue:                        workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), controllerName),
+		upstreamDynamicClusterClient: upstreamDynamicClusterClient,
+		downstreamDynamicClient:      downstreamDynamicClient,
+		upstreamEventHandlers:        []ResourceEventHandlerPerGVR{},
+		downstreamEventHandlers:      []ResourceEventHandlerPerGVR{},
+		syncerInformerMap:            map[schema.GroupVersionResource]*SyncerInformer{},
+		syncTargetName:               syncTargetName,
+		syncTargetWorkspace:          syncTargetWorkspace,
+		syncTargetUID:                syncTargetUID,
+		syncTargetLister:             syncTargetInformer.Lister(),
+	}
+
+	logger := logging.WithReconciler(klog.Background(), controllerName)
+
+	syncTargetInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: func(obj interface{}) bool {
+			key, err := kcpcache.DeletionHandlingMetaClusterNamespaceKeyFunc(obj)
+			if err != nil {
+				return false
+			}
+			clusterName, _, name, err := kcpcache.SplitMetaClusterNamespaceKey(key)
+			if err != nil {
+				return false
+			}
+			return name == syncTargetName && clusterName == syncTargetWorkspace
+		},
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) { c.enqueueSyncTarget(obj, logger) },
+			UpdateFunc: func(old, obj interface{}) {
+				oldCluster := old.(*workloadv1alpha1.SyncTarget)
+				newCluster := obj.(*workloadv1alpha1.SyncTarget)
+
+				// only enqueue when syncedResource is changed.
```

Comment:
fundamentally is it ok to do this? why? why not always enqueue? what if we miss an event?

Reply:
fixed